MediaBrowser.Providers/MediaInfo/AudioResolver.cs

```diff
@@ -0,0 +1,275 @@
+using MediaBrowser.Controller.Providers;
+using MediaBrowser.Model.Dlna;
+using MediaBrowser.Model.Dto;
+using MediaBrowser.Model.Entities;
+using MediaBrowser.Model.Globalization;
+using MediaBrowser.Model.MediaInfo;
+
+namespace MediaBrowser.Providers.MediaInfo
+{
+    public class AudioResolver
+    {
+        private readonly ILocalizationManager _localization;
+
+        private readonly IMediaEncoder _mediaEncoder;
+
+        private readonly CancellationToken _cancellationToken;
+
+        public AudioResolver(ILocalizationManager localization, IMediaEncoder mediaEncoder, CancellationToken cancellationToken = default)
+        {
+            _localization = localization;
+            _mediaEncoder = mediaEncoder;
+            _cancellationToken = cancellationToken;
+        }
+
+        public List<MediaStream> GetExternalAudioStreams(
+            Video video,
+            int startIndex,
+            IDirectoryService directoryService,
+            bool clearCache)
+        {
+            var streams = new List<MediaStream>();
+
+            if (!video.IsFileProtocol)
+            {
+                return streams;
+            }
+
+            AddExternalAudioStreams(streams, video.ContainingFolderPath, video.Path, startIndex, directoryService, clearCache);
+
+            startIndex += streams.Count;
+
+            string folder = video.GetInternalMetadataPath();
+
+            if (!Directory.Exists(folder))
+            {
+                return streams;
+            }
+
+            try
+            {
+                AddExternalAudioStreams(streams, folder, video.Path, startIndex, directoryService, clearCache);
+            }
+            catch (IOException)
+            {
+            }
+
+            return streams;
+        }
+
+        public IEnumerable<string> GetExternalAudioFiles(
+            Video video,
+            IDirectoryService directoryService,
+            bool clearCache)
+        {
+            if (!video.IsFileProtocol)
+            {
+                yield break;
+            }
+
+            var streams = GetExternalAudioStreams(video, 0, directoryService, clearCache);
+
+            foreach (var stream in streams)
+            {
+                yield return stream.Path;
+            }
+        }
+
+        public void AddExternalAudioStreams(
+            List<MediaStream> streams,
+            string videoPath,
+            int startIndex,
+            IReadOnlyList<string> files)
+        {
+            var videoFileNameWithoutExtension = NormalizeFilenameForAudioComparison(videoPath);
+
+            for (var i = 0; i < files.Count; i++)
+            {
+
+                var fullName = files[i];
+                var extension = Path.GetExtension(fullName.AsSpan());
+                if (!IsAudioExtension(extension))
+                {
+                    continue;
+                }
+
+                Model.MediaInfo.MediaInfo mediaInfo = GetMediaInfo(fullName).Result;
```

Comment:
I think I need some help here. I tried with async & await but I can't declare the method as async because some variables are ReadonlySpans which aren't allowed in async methods.

Reply:
So don't use Spans :) It's better to use strings and async than using spans and sync.